pkg/dns-server/resolver/server_test.go

```diff
@@ -0,0 +1,202 @@
+
+		_, err = resolver.AddServiceToDomain("service", "kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		ip, err := resolver.ForwardLookup("service.kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		stop := make(chan struct{})
+		go func() {
+			err := resolver.Start(stop)
+			Expect(err).ToNot(HaveOccurred())
+		}()
+
+		client := new(dns.Client)
+		message := new(dns.Msg)
+		_ = message.SetQuestion("service.kuma.", dns.TypeA)
+		var response *dns.Msg
+		Eventually(func() error {
+			response, _, err = client.Exchange(message, "127.0.0.1:5653")
+			return err
+		}).ShouldNot(HaveOccurred())
+		Expect(response.Answer[0].String()).To(Equal(fmt.Sprintf("service.kuma.\t3600\tIN\tA\t%s", ip)))
+
+		message = new(dns.Msg)
+		_ = message.SetQuestion("backend.kuma.", dns.TypeA)
+		response, _, err = client.Exchange(message, "127.0.0.1:5653")
+		Expect(err).ToNot(HaveOccurred())
+		Expect(len(response.Answer)).To(Equal(0))
+
+		// stop the resolver
+		stop <- struct{}{}
+
+		// ready
+		close(done)
+	})
+
+	It("DNS Server domain operation", func(done Done) {
+		resolver, err := NewSimpleDNSResolver("127.0.0.1", "5653", "240.0.0.0/4")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain("")
+		Expect(err).To(HaveOccurred())
+
+		err = resolver.AddDomain("kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain(".kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain(".other")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain(".third")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain(".kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.RemoveDomain(".kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.RemoveDomain(".other")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.RemoveDomain(".other")
+		Expect(err).To(HaveOccurred())
+
+		// ready
+		close(done)
+	})
+
+	It("DNS Server service operation", func(done Done) {
+		resolver, err := NewSimpleDNSResolver("127.0.0.1", "5653", "240.0.0.0/4")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain("kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain("other")
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = resolver.AddServiceToDomain("service", "third")
+		Expect(err).To(HaveOccurred())
+
+		_, err = resolver.AddServiceToDomain("service", "kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = resolver.AddServiceToDomain("backend", "kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = resolver.AddServiceToDomain("service", "kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.RemoveServiceFromDomain("service", "kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.RemoveServiceFromDomain("service", "kuma")
+		Expect(err).To(HaveOccurred())
+
+		err = resolver.RemoveServiceFromDomain("service", "third")
+		Expect(err).To(HaveOccurred())
+
+		// ready
+		close(done)
+	})
+
+	It("DNS Server sync operation", func(done Done) {
```

Comment:
I think this test needs more structure (given, when, then. Multiple Its). I'm happy to pair program to refactor it